### PR TITLE
Allow multiple setTrailer

### LINKED ISSUE #854: SetTrailer should be callable multiple times
Making [`SetTrailer`](https://github.com/grpc/grpc-go/blob/0032a855/server.go#L883) callable only once from a unary RPC handler limits the utility of server-side interceptors. Specifically, we would like to be able to add a trailer to return server health on every RPC call, so that the health status can be updated in clients. If we do that now, we have to disallow all other use of trailers by unary RPC handlers.

cc @maniksurtani @jhump @lukaszx0

### PR DESCRIPTION
**Pull request by menghanl**
fixes #854 
updates #733

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -880,8 +880,9 @@ func SendHeader(ctx context.Context, md metadata.MD) error {
 }
 
 // SetTrailer sets the trailer metadata that will be sent when an RPC returns.
-// It may be called at most once from a unary RPC handler. The ctx is the RPC
-// handler's Context or one derived from it.
+// It may be called multiple times from a unary RPC handler.
+// When called more than once, all the provided metadata will be merged.
```
Comment (zellyn) [inline server.go]: s/merged/merged by appending/ or something explaining how the merging works.

Comment (menghanl) [inline server.go]: How merging works is not important. The point is that all metadata will be sent, not just the last one.
What's the thing you find confusing here? If you think "merged" is not a good word, what about `all md's will be combined`? Or other suggestions?

Comment (zellyn) [inline server.go]: Until I read the code and found the slice append, I was worried that it would "merge" maps by overwriting existing keys.

**Comment by NeoCN** [comment]
looking forward to see this pull request being merged

File: test/end2end_test.go [hunk 2]
```diff
@@ -1193,6 +1208,97 @@ func testMetadataUnaryRPC(t *testing.T, e env) {
...
+}
+
+func testMultipleSetTrailerUnaryRPC(t *testing.T, e env) {
+	te := newTest(t, e)
+	te.startServer(&testServer{security: e.security, multipleTrailer: true})
+	defer te.tearDown()
+	tc := testpb.NewTestServiceClient(te.clientConn())
+
+	const argSize = 1
+	const respSize = 1
```
Comment (iamqizhao) [inline test/end2end_test.go]:
```
const (
  argSize =  1
  respSize = 1
)
```

File: test/end2end_test.go [hunk 3]
```diff
@@ -1193,6 +1208,97 @@ func testMetadataUnaryRPC(t *testing.T, e env) {
...
+	}
+
+	req := &testpb.SimpleRequest{
+		ResponseType: testpb.PayloadType_COMPRESSABLE.Enum(),
+		ResponseSize: proto.Int32(respSize),
+		Payload:      payload,
+	}
+	var trailer metadata.MD
+	ctx := metadata.NewContext(context.Background(), testMetadata)
+	if _, err := tc.UnaryCall(ctx, req, grpc.Trailer(&trailer)); err != nil {
```
Comment (iamqizhao) [inline test/end2end_test.go]: set it to non-failfast?

**Pull request merged**